### PR TITLE
Showing warning on FixedEditText instead of Toast

### LINKED ISSUE #13631: [BUG]: Show warning on FixedEditText instead of Toast
### Checked for duplicates?

- [X] This issue is not a duplicate

### What are the steps to reproduce this bug?

1. Click on `manage note types` from DeckPicker.kt
2. Select any one of the note type from the list
3. Add new field by clicking `+` from top right corner
4. Click OK without entering value

### Expected behaviour

The warning should show on FixedEditText when user clicks on OK without entering any value

### Actual behaviour

Showing Toast when user clicks on OK without entering any value and Dialog is closed

### Debug info

```text
AnkiDroid Version = 2.16alpha97-debug

Android Version = 13

ProductFlavor = amazon

Manufacturer = realme

Model = RMX3521

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 13; RMX3521 Build/RKQ1.211119.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/111.0.5563.116 Mobile Safari/537.36

ACRA UUID = c070f6dc-866e-4fcf-8bc1-2aa53bcca93d

New schema = false

Scheduler = std2

Crash Reports Enabled = false

DatabaseV2 Enabled = true
```

### (Optional) Anything else you want to share?

_No response_

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by sanjaysargam**
## Purpose / Description
It's better to show a warning on FixedEditText rather than Toast 

## Fixes
Fixes #13631

## Approach
Implementing a warning message on FixedEditText when user didn't enter anything in FixedEditText or the value is duplicate

## How Has This Been Tested?

Tested on Physical Device (Realme 9)

https://user-images.githubusercontent.com/65113071/233145117-79d5e4a0-1e05-429a-a026-87531f170ee9.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — COMMENTED** [review]
Thanks!

One potential issue, one further UX improvement to look into

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 1]
```diff
@@ -194,9 +194,11 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
                     }
                     collection.models.update(mModel)
                     initialize()
+                    this.dismiss()
                 }
                 negativeButton(R.string.dialog_cancel)
             }
+                .noAutoDismiss()
```
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:201]: Ohh forgot to dismiss dialog in negative button
I'll do

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 2]
```diff
@@ -173,6 +165,14 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
                 positiveButton(R.string.dialog_ok) {
                     // Name is valid, now field is added
                     val fieldName = uniqueName(_fieldNameInput)
+                    if (fieldName!!.isEmpty()) {
+                        _fieldNameInput.error = getString(R.string.toast_empty_name)
+                        return@positiveButton
+                    }
+                    if (mFieldsLabels.any { fieldName == it }) {
+                        _fieldNameInput.error = getString(R.string.toast_duplicate_field)
```
[the remark this answers is not in the document]
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:173]: Yeah!! That's better approach
will implement this

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 3]
```diff
@@ -166,13 +160,31 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
...
         fieldNameInput?.let { _fieldNameInput ->
+            // Add a text change listener to show a warning for duplicate
+            val textWatcher = object : TextWatcher {
+                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
+
+                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+                    val fieldName = s.toString().trim()
+                    if (mFieldsLabels.any { fieldName == it }) {
+                        _fieldNameInput.error = getString(R.string.toast_duplicate_field)
+                    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:171]: We should disable the positive button if an error occurred.

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 4]
```diff
@@ -166,13 +160,31 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
...
             MaterialDialog(this).show {
                 customView(view = _fieldNameInput, horizontalPadding = true)
                 title(R.string.model_field_editor_add)
                 positiveButton(R.string.dialog_ok) {
                     // Name is valid, now field is added
                     val fieldName = uniqueName(_fieldNameInput)
+                    if (fieldName!!.isEmpty()) {
+                        _fieldNameInput.error = getString(R.string.toast_empty_name)
+                        return@positiveButton
+                    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:187]: Previously: `uniqueName` performed two types of validation, now it only performs one

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 5]
```diff
@@ -288,16 +304,30 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
...
+                    if (mFieldsLabels.any { fieldName == it }) {
+                        _fieldNameInput.error = getString(R.string.toast_duplicate_field)
+                    }
+                    if (fieldName.isEmpty()) {
+                        _fieldNameInput.error = getString(R.string.toast_empty_name)
+                    }
+                }
+
+                override fun afterTextChanged(s: Editable?) {}
+            }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:322]: This seems to be duplicated code, see if it can be combined into a single class

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt (hunk 5 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:322]: In which file, this is need to be added

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt (hunk 4 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:187]: uniqueName function checks two condition 
1. whether the input is empty or not
2. whether the input is already exists or not

These two validation, I already added

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 6]
```diff
@@ -79,6 +85,13 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
             subtitle = intent.getStringExtra("title")
         }
         startLoadingCollection()
+        textWatcher = object : TextWatcher {
+            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
+
+            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
+
+            override fun afterTextChanged(s: Editable?) {}
+        }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:94]: Could you explain this in a comment?

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 7]
```diff
@@ -291,13 +313,23 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
...
+                    val fieldName = text.toString().trim()
+                    if (mFieldsLabels.any { fieldName == it }) {
+                        _fieldNameInput.error = getString(R.string.toast_duplicate_field)
+                        getActionButton(WhichButton.POSITIVE).isEnabled = false
+                    } else if (fieldName.isEmpty()) {
+                        _fieldNameInput.error = getString(R.string.toast_empty_name)
+                        getActionButton(WhichButton.POSITIVE).isEnabled = false
+                    } else {
+                        getActionButton(WhichButton.POSITIVE).isEnabled = true
                     }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:330]: Add a TODO for this functionality to be extracted

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt [hunk 8]
```diff
@@ -194,9 +212,13 @@ class ModelFieldEditor : AnkiActivity(), LocaleSelectionDialogHandler {
...
                     collection.models.update(mModel)
                     initialize()
+                    this.dismiss()
+                }
+                negativeButton(R.string.dialog_cancel) {
+                    this.dismiss()
                 }
-                negativeButton(R.string.dialog_cancel)
             }
+                .noAutoDismiss()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:221]: Why does this exist if the positive button is disabled when the input is invalid and both buttons dismiss the dialog?

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt (hunk 6 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:94]: Ok will add comment

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt (hunk 7 above)
Comment (SanjaySargam) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.kt:330]: Did you mean to add TODO in comment?

**Comment by SanjaySargam** [comment]
Hey @david-allison Please open this PR

**Pull request closed without merging**